File: 0671-second-minimum-node-in-a-binary-tree/0671-second-minimum-node-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        long long a=LLONG_MAX;
        long long b=LLONG_MAX;
        fun(root,a,b);
        if(b==LLONG_MAX) return -1;
        return (int)b;
    }
private :
    void fun(TreeNode* root,long long& a,long long& b){
        if(root==nullptr) return;
        if(root->val<a) a=root->val;
        if(root->val>a && root->val<b) b=root->val;
        fun(root->left,a,b);
        fun(root->right,a,b);
    }
};

File: 0671-second-minimum-node-in-a-binary-tree/0671-second-minimum-node-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findSecondMinimumValue(TreeNode root) {
        long[] a=new long[1];
        long[] b=new long[1];
        a[0]=Long.MAX_VALUE;
        b[0]=Long.MAX_VALUE;
        fun(root,a,b);
        if(b[0]==Long.MAX_VALUE) return -1;
        return (int)b[0];
    }
    private void fun(TreeNode root,long[] a,long[] b){
        if(root==null) return;
        if(root.val<a[0]) a[0]=root.val;
        if(root.val>a[0] && root.val<b[0]) b[0]=root.val;
        fun(root.left,a,b);
        fun(root.right,a,b);
    }
}